audiostream/src/ni/media/audio/source/gstreamer_file_source.cpp

```diff
@@ -0,0 +1,337 @@
+{
+    if ( format == 'U' )
+        return pcm::unsigned_integer;
+    else if ( format == 'F' )
+        return pcm::floating_point;
+
+    return pcm::signed_integer;
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::onPadAdded( GstElement* element, GstPad* pad, GstElement* sink )
+{
+    tGstPtr<GstPad> sinkpad( gst_element_get_static_pad( sink, "sink" ), gst_object_unref );
+    gst_pad_link( pad, sinkpad.get() );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+std::streampos gstreamer_file_source::seek( offset_type off, BOOST_IOS::seekdir way )
+{
+    assert( 0 == off % m_info.bytes_per_frame() );
+
+    int64_t newPosition = 0;
+
+    if ( way == BOOST_IOS::seekdir::_S_beg )
+    {
+        newPosition = off;
+    }
+    else if ( way == BOOST_IOS::seekdir::_S_cur )
+    {
+        newPosition = m_position + off;
+    }
+    else if ( way == BOOST_IOS::seekdir::_S_end )
+    {
+        int64_t end = 0;
+        if ( !gst_element_query_duration( m_pipeline.get(), GST_FORMAT_BYTES, &end ) )
+        {
+            throw std::runtime_error(
+                "gstreamer_file_source: seeking from end of file impossible - could not query duration" );
+        }
+        newPosition = end + off;
+    }
+
+    if ( m_position != newPosition )
+    {
+        m_position = newPosition;
+        m_ring_buffer->flush();
+
+        if ( !gst_element_seek_simple( m_pipeline.get(),
+                                       GST_FORMAT_BYTES,
+                                       ( GstSeekFlags )( GST_SEEK_FLAG_FLUSH | GST_SEEK_FLAG_ACCURATE ),
+                                       m_position ) )
+        {
+            throw std::runtime_error( "gstreamer_file_source: seeking failed" );
+        }
+    }
+
+    return m_position;
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+std::streamsize gstreamer_file_source::read( char* dst, std::streamsize numBytesRequested )
+{
+    return recursive_read( dst, numBytesRequested );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+std::streamsize gstreamer_file_source::recursive_read( char* dst, std::streamsize numBytesRequested )
+{
+    auto read_bytes = m_ring_buffer->pull( dst, numBytesRequested );
+
+    dst += read_bytes;
+    numBytesRequested -= read_bytes;
+    m_position += read_bytes;
+
+    if ( numBytesRequested )
+    {
+        tGstPtr<GstElement> sink( gst_bin_get_by_name( GST_BIN( m_pipeline.get() ), "sink" ), gst_object_unref );
+        GstAppSink*         app_sink = reinterpret_cast<GstAppSink*>( sink.get() );
+
+        tGstPtr<GstSample> sample( gst_app_sink_pull_sample( app_sink ), (GUnref) gst_sample_unref );
```

Comment:
This will block until a buffer is available, and you would not notice any errors that happen in the meantime and block here forever. See my general comment about the main loop